_posts/2025-05-09-sorting-for-fast-selective-queries.md

```diff
@@ -0,0 +1,660 @@
+Creative sorting approaches that use space filling curves and/or rounded timestamp columns allow this approach to work effectively across multiple columns.
+Once your dataset becomes large or you are storing it remotely, consider applying these techniques.
+Plus, these approaches can be used in nearly any columnar file format or database!
+
+Happy analyzing!
+
+## Appendix: Experiment Details
+
+All experiments were run on an M1 MacBook Pro with DuckDB v1.2.2.
+Note that the remote S3 tests were run over WiFi (which adds a bit of variability...), so please feel free to benchmark independently!
+
+### Table Creation Time
+
+Creating a DuckDB table from Parquet files without sorting took slightly over 21 seconds.
+Each other approach copied from the unsorted DuckDB table and created a new table.
+The various methods of sorting required similar amounts of time (between 48 and 61 seconds), so we are free to choose the one with the most effective results without considering relative insert performance.
+However, it is worth noting that overall insert performance slows down by nearly 3× with any sorting.
+
+| Table name       | Creation time (s) |
+| :--------------- | ----------------: |
+| `from_parquet`   |              21.4 |
+| `random`         |              60.2 |
+| `origin`         |              51.9 |
+| `origin_dest`    |              48.7 |
+| `zipped_varchar` |              58.2 |
+| `morton`         |              54.6 |
+| `hilbert`        |              58.5 |
+| `hilbert_day`    |              58.7 |
+| `hilbert_month`  |              53.8 |
+| `hilbert_year`   |              60.2 |
+
+### Basic Sorting
+
+<details markdown='1'>
+<summary markdown='span'>
+    For reproducibility, here are the very standard queries used to initially load the data from Parquet, sort randomly, sort by `origin`, and sort by `origin` and then `destination`.
+</summary>
+
+```sql
+CREATE TABLE IF NOT EXISTS flights AS
+    FROM './Combined_Flights*.parquet';
+
+CREATE TABLE IF NOT EXISTS flights_random AS
+    FROM flights
+    ORDER BY
+        hash(rowid + 42);
+
+CREATE TABLE IF NOT EXISTS flights_origin AS
+    FROM flights
+    ORDER BY
+        origin;
+
+CREATE TABLE IF NOT EXISTS flights_origin_dest AS
+    FROM flights
+    ORDER BY
+        origin,
+        dest;
+```
+
+</details>
+
+## Sorting by Zipped `VARCHAR` Columns
+
+As an example of an approach that does not require an extension, this SQL macro roughly approximates a space filling curve approach, but using alphanumeric characters instead of integers.
+The outcome is a dataset that is somewhat sorted by one column and somewhat sorted by another.
+
+```sql
+CREATE OR REPLACE FUNCTION main.zip_varchar(i, j, num_chars := 6) AS (
+    -- By default using 6 characters from each string so that
+    -- if data is ASCII, we can fit it all in the 12 byte inline portion of DuckDB's string representation
+    [
```

Comment:
I rewrote this with function chaining to make it more readable. Please double-check!

Reply:
Super close! Just one extra space in the `array_to_string(' ')`. Thanks for the change - it is cleaner!